src/version_check.py

```diff
@@ -0,0 +1,231 @@
+        """Returns a int for the version of the related app.
+
+        Raises:
+            NoVersionError.
+        """
+        try:
+            for relation_name in self.relations_to_check:
+                for rel in self.charm.model.relations[relation_name]:
+                    if rel.app.name == related_app_name:
+                        return int(rel.data[rel.app][VERSION_CONST])
+        except KeyError:
+            pass
+
+        raise NoVersionError(f"Expected {related_app_name} to have version info.")
+
+    def get_deployment_prefix(self) -> str:
+        """Returns the deployment prefix, indicating if the charm is locally deployred or not.
+
+        TODO: Keep this until ops framework supports:
+            https://github.com/canonical/operator/issues/1255
+        """
+        file_path = f"{PREFIX_DIR}/unit-{self.charm.unit.name.replace('/','-')}/charm/.juju-charm"
+        with open(file_path) as f:
+            charm_path = f.read().rstrip()
+
+        return charm_path.split(":")[0]
+
+    def are_related_apps_valid(self) -> bool:
+        """Returns True if a related app has a version that's incompatible with the current app.
+
+        Raises:
+            NoVersionError.
+        """
+        return self.get_invalid_versions() == []
+
+    def set_version_across_all_relations(self) -> None:
+        """Sets the version number across all related apps, prvided by relations_to_check."""
+        if not self.charm.unit.is_leader():
+            return
+
+        for relation_name in self.relations_to_check:
+            for rel in self.charm.model.relations[relation_name]:
+                rel.data[self.charm.model.app][VERSION_CONST] = str(self.version)
+                rel.data[self.charm.model.app][DEPLOYMENT_TYPE] = str(self.get_deployment_prefix())
+
+    def set_version_on_related_app(self, relation_name: str, related_app_name: str) -> None:
+        """Sets the version number across for a specified relation on a specified app."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relations = self.charm.model.relations[relation_name]
+        for rel in relations:
+            if rel.app.name == related_app_name:
+                rel.data[self.charm.model.app][VERSION_CONST] = str(self.version)
+                rel.data[self.charm.model.app][DEPLOYMENT_TYPE] = str(self.get_deployment_prefix())
+
+    def set_version_on_relation_created(self, event) -> None:
+        """Shares the charm's revision to the newly integrated application.
+
+        Raises:
+            RelationInvalidError
+        """
+        if event.relation.name not in self.relations_to_check:
+            raise RelationInvalidError(
+                f"Provided relation: {event.relation.name} not in self.relations_to_check."
+            )
+
+        self.set_version_on_related_app(event.relation.name, event.app.name)
+
+    def is_local_charm(self, app_name: str) -> bool:
```

Comment:
@Mehdi-Bendriss this function is not *yet* used in the charm, but we could use it. 

During a chat with Billy Olsen yesterday, he mentioned a potential scenario of the support team patching a hot fix to the charm. In this case the hotfix would result in the charm having a localhost revision. This would result in the relations showing either:

shard - Blocked(Charm revision 0 is not up-to date with config-server (Y))
config-server - Waiting(Waiting for all cluster components to upgrade to revision 0)

and the support teams fix not working. 

What we could do is if there is a revision mismatch and one of the charms is a local charm, we could not go into Blocked or Waiting, and instead log a `warning` with something like `WARNING: deploying a local charm, cannot check revision across components.` Should we do this?

Reply:
I think that's a sensible approach

Reply:
yes, seems reasonable